checker/src/org/checkerframework/checker/index/upperbound/UpperBoundStore.java

```diff
@@ -0,0 +1,348 @@
+                    if (!((FlowExpressions.FieldAccess) r).getReceiver().toString().equals("this")
+                            && !((FlowExpressions.FieldAccess) r).isFinal()) {
+                        if (checkController == CheckController.NONFINAL_REFS_AND_METHOD_CALLS) {
+                            return Result.failure(NO_REASSIGN_FIELD, anno.toString());
+                        } else {
+                            return Result.failure(SIDE_EFFECTING_METHOD, anno.toString());
+                        }
+                    } else {
+                        FlowExpressions.Receiver oldR = r;
+                        try {
+                            r =
+                                    factory.getReceiverFromJavaExpressionString(
+                                            ((FlowExpressions.FieldAccess) r).getField().toString(),
+                                            factory.getPath(n.getTree()));
+                        } catch (FlowExpressionParseUtil.FlowExpressionParseException e) {
+                        }
+                        if (oldR.equals(r)) {
+                            break;
+                        }
+                    }
+                } else {
+                    break;
+                }
+            }
+            if (r instanceof FlowExpressions.MethodCall) {
+                if (checkController == CheckController.NAME_AND_METHOD_CALLS
+                        || checkController == CheckController.NONFINAL_REFS_AND_METHOD_CALLS) {
+                    return Result.failure(NO_REASSIGN_FIELD_METHOD, anno.toString());
+                }
+                if (checkController == CheckController.NONFINAL_REFS_AND_METHOD_CALLS_METHOD) {
+                    return Result.failure(SIDE_EFFECTING_METHOD, anno.toString());
+                }
+                for (FlowExpressions.Receiver param :
+                        ((FlowExpressions.MethodCall) r).getParameters()) {
+                    canonicalDependencies.add(param.toString());
+                }
+            }
+        }
+        if ((checkController == CheckController.NAME_AND_METHOD_CALLS
+                        || checkController == CheckController.NAME_ONLY)
+                && canonicalDependencies.contains(canonicalTargetName)) {
+            return Result.failure(
+                    NO_REASSIGN, canonicalTargetName, anno.toString(), canonicalTargetName);
+        }
+        return Result.SUCCESS;
+    }
+
+    void checkForRemainingAnnotations(
+            List<AnnotatedTypeMirror> enclosedTypes,
+            String canonicalTargetName,
+            Node n,
+            CheckController checkController) {
+        for (AnnotatedTypeMirror atm : enclosedTypes) {
+            for (AnnotationMirror anno : atm.getAnnotations()) {
+                Result r = checkAnno(anno, canonicalTargetName, n, checkController);
+                if (r.isFailure()) {
+                    checker.report(r, n.getTree());
```

Comment:
I thought about that and couldn't see any way to do it, but you know the codebase better than I do. We should revisit this issue once we've improved this code.